### PR TITLE
Problem: name confl from windows.h, draft API decl in pch.

### PR DESCRIPTION
**Pull request by evoskuil**
See https://github.com/zeromq/libzmq/issues/1989

### CODE REVIEW

File: src/proxy.cpp [hunk 1]
```diff
@@ -46,8 +46,9 @@
 #include "socket_base.hpp"
 #include "err.hpp"
 
+// TODO: determine if this is an issue, since zmq.h is being loaded from pch.
 // zmq.h must be included *after* poll.h for AIX to build properly
```
Comment (evoskuil) [inline src/proxy.cpp]: I wanted to highlight this comment. The previous implementation did not conform to the requirement indicated by the comment. I don't have the ability to test this, but this pull request does not change the actual behavior.

File: src/zmq.cpp [hunk 2]
```diff
@@ -40,16 +40,14 @@
 #include <poll.h>
 #endif
 
+// TODO: determine if this is an issue, since zmq.h is being loaded from pch.
 // zmq.h must be included *after* poll.h for AIX to build properly
```
Comment (evoskuil) [inline src/zmq.cpp]: I wanted to highlight this comment (also). The previous implementation did not conform to the requirement indicated by the comment. I don't have the ability to test this, but this pull request does not change the actual behavior.

**Comment by bluca** [comment]
Note that the precompiled headers support is only available on Windows.
This means that on *nix,  by removing the include of platform.hpp, we are at the mercy of the preprocessor - if it happens to parse first a file that includes the platform.hpp before it includes this headers, all the constants below will be available, else they won't.
So we should either leave platform.hpp included where it is, or manually include precompiled.hpp everywhere you are removing the include of platform.hpp.

**Comment by evoskuil** [comment]
> Note that the precompiled headers support is only available on Windows.

@bluca The only aspect of the implementation that pertains to precompilation is these instructions within the msvc `libzmq.props` and `libzmq.vcxproj` files:

```xml
<ClCompile>
    <PrecompiledHeader>Use</PrecompiledHeader>
    <PrecompiledHeaderFile>precompiled.hpp</PrecompiledHeaderFile>
    ...
</ClCompile>
```

```xml
<ClCompile Include="..\..\..\..\src\precompiled.cpp">
    <PrecompiledHeader>Create</PrecompiledHeader>
</ClCompile>
...
<ClCompile Include="..\..\..\..\src\tweetnacl.c">
    <PrecompiledHeader>NotUsing</PrecompiledHeader>
</ClCompile>
```

So it doesn't matter whether another compiler supports or is configured to apply precompiled headers.

> This means that on *nix, by removing the include of platform.hpp, we are at the mercy of the preprocessor...

The include of the `platform.hpp` isn't removed, it's moved to the `precompiled.hpp`. What I didn't notice is that `precompiled.cpp` is not included in the *nix makefile. I did not expect this and will correct it.

There is no harm if for some reason precompilation is not going to be used with the other compilers. And if the name "precompiled" is disfavored because precompilation isn't going to be configured for other compilers, then the name of the file can certainly be changed.

The preceding design produces namespace conflicts arising from the fact that it is impossible to see or control when the `windows.hpp` will be included. As a result a slight change to the project can cause it to be included after some other header that is intended to be affected by the configuration in `windows.hpp`. It is essentially a requirement that in every file headers are included in the following order:

* `precompiled.hpp`
* `platform.hpp`
* `windows.hpp` (as applicable).
* ...

If we don't do the above, we are at the mercy of the compiler's include ordering - which created a number of errors in the Windows build when I added the `zmq_draft.hpp` - which should not have been a problem.

> or manually include precompiled.hpp everywhere you are removing the include of platform.hpp.

I understood it to be a requirement of the preceding design that `precompiled.hpp` must be the first include of every cpp file, so I did not consider this a change.

**Comment by bluca** [comment]
If it's fixed in autotools/cmake then it's fine by me as a solution - given the platform.hpp is a result of the build system generating a configuration, it's ok IMHO to have it included by the same build system that creates it.

**Comment by bluca** [comment]
@evoskuil - don't forget CMake :-)

**Comment by evoskuil** [comment]
The above commit will resolve any potential problem of `platform.hpp` not having precedence in any build, though I'm anticipating an empty compilation unit warning.

**Comment by evoskuil** [comment]
@bluca Okay, I didn't realize that CMake had it's own file listing - thanks.

**Comment by evoskuil** [comment]
@bluca Turns out that `precompiled.cpp` is already included by CMake, so it was only missing in the autotools builds.

**Comment by bluca** [comment]
Ah sorry should have checked, thanks for looking into it!

**Comment by evoskuil** [comment]
There are no empty compilation unit warnings, so I guess those are already suppressed.

**Comment by bluca** [comment]
Perfect. Looks good to me - do you have any more pending changes? Else I'll go ahead and merge

**Comment by evoskuil** [comment]
@bluca I've got nothing queued up, all done here.

**Comment by evoskuil** [comment]
@bluca Thanks for the help!

**Pull request merged**